### ISSUE TITLE
Add well-defined env vars to launcher/device environments

### ISSUE THREAD

**Issue by knolleary** [flowforge/flowforge#841]
### Description

**As a:** flow developer

**I want to:** access to project properties in my flows

**So that:** I can create a project that deploys to multiple devices without having to manually set per-device env vars

--

In summary, I propose we add the following set of env vars to the default launcher and device environments:

 - `FF_PROJECT_ID`
 - `FF_PROJECT_NAME`
 - `FF_DEVICE_ID`
 - `FF_DEVICE_NAME`

### Which customers would this be availble to

All Users, (CE)

**Comment by Steve-Mcl** [comment]
Current thoughts and questions...
1. Should these be implicit (added to env behind the scenes from the launcher) or explicit (added in the FF UI so users know they are there?)
   1.  If visible in the UI for the project/device, they should be locked/uneditble
   2. Should the value be shown or starred?
1. If whilst implementing and testing, it becomes clear more variables would make sense (even if they are edge case), should I add them for review or discuss here first?
1. Should project nodes be considered here (i.e. add additional info to the msg about the source project/device) or is that out of scope/covered in another task?

**Comment by knolleary** [comment]
1. Initially they can be implicit. There are UX complications around how they would be displayed. For example, all of the device ones will be specific to the device the project is deployed to - so it isn't obvious what the UI would show in the project view of the env vars.
2. I've listed the ones that need to be added. Any others should be discussed, either in the issue or in the PR review.
3. Out of scope of this story. If there is a need to add additional meta data to the project node messages, then a story should be raised against the project nodes and it get prioritised as usual.

**Comment by Steve-Mcl** [comment]
One more question, should these new env vars be allowed to be set/overridden by the user explicitly by entering same named variables? Or should we prohibit entry of these "known" env vars?

I feel there are arguments for both. So I would like to be explicit about the implementation so I can add suitable tests. Ta.

**Comment by knolleary** [comment]
They are platform defined and must be predictable. The user must not be able to override them.

**Comment by Steve-Mcl** [comment]
So, this seemed straight forward until I come to implement.

Am i right in thinking the device-agent will need modifying to include new env-vars? And that these are passed in via the `device.yml`?

![Image](https://user-images.githubusercontent.com/44235289/193588609-44960b64-86b1-42cd-ba7c-4855370aeabc.png)

QUESTION: If we include `FF_PROJECT_NAME` and `FF_DEVICE_NAME` in this manor, then they will be set and loaded into device memory until the credentials are regenerated - what if the admin renames the project or renames the device?

**Comment by knolleary** [comment]
No. Env Vars are retrieved as part of the `/api/v1/project/:deviceId/live/settings` end point - https://github.com/flowforge/flowforge/blob/8d4f374e6662f2492dd33fa655a9221032c2d64e/forge/routes/api/deviceLive.js#L97 - that is called whenever the agent restarts node-red.

**Comment by Steve-Mcl** [comment]
Another question...

We can add env vars with or without re-computing the settings hash.
* Without recomputing the settings has, the device launcher does not know there are new env vars
* With recomputing the settings has, the device launcher sees a change and restarts the instance

The risks are: running with out of step env vars VS restarting all devices because env vars were updated.


This leads to other points of interest...
* If we intend on recalculating the hash, then we might as well add them to the `settings.env` property - which means we _CAN_ display them in the FF env settings (i think)

**Comment by knolleary** [comment]
There are two different things to consider because this feature isn't just about devices:

1. How are Project-specific generated EVs handled
2. How are Device-specific generated EVs handled

The project EVs should be captured in the snapshot.

The settings hash should reflect the total set of Device-specific EVs - both user- and platform-defined. Otherwise, as you say, the device won't know when a value has changed. That said, the only platform-defined EV we currently have planned that *might* change is `DEVICE_NAME`.

If they get stored in the DB and shown to the user, then you need a way in the UI to identify them as platform-defined so that the UI doesn't allow them to be edited (likewise, the API needs to ensure the user cannot manually defined an EV with a 'reserved' name.

I don't have a strong view on whether they show get stored in the db alongside the other EVs, or dynamically generated when needed - I can see benefits either way.

**Comment by Steve-Mcl** [comment]
Draft PR raised early to ratify direction of work. #1039 

Questions...
* Should we have similar EVs in the project level?
  * What comes to mind is we could / should move the project_id and project_name to the project (since the device inherits these)
  * This then begs the question should inherited EVs be visible & decernible in the device 🤔 

### Progress...

1. Have platform EVs added dynamically (not saved to DB)
2. Have device `name` and `type` also update settings hash (and therefore instruct device to reload settings)
   a. Since device name and type are present in the EVs, this PR also signals the setting change to the device
3. Since it was more work to maintain separate EVs for UI and for Device, I left the system EVs present in the UI and added a Cog to indicate system type EV

![image](https://user-images.githubusercontent.com/44235289/193874266-83a95338-79c4-4963-8b0f-1003d5f31c95.png)

### Outstanding
* Depending on answers to above questions

### TODO
* Add tests

**Comment by knolleary** [comment]
> Are we ok with FF_SNAPSHOT_ID/NAME being included

Not sure. Given how snapshots may evolve in the future, I wouldn't necessarily want to expose them like this yet - particularly without a concrete use case.

For the purposes of this isue - getting the basic infrastructure in place to have these platform provided EVs, can we stick to the ones listed in the issue? Otherwise the is lots of discussion about other EVs that, once the infrastructure is in place, should take moments to add in a follow up PR.

> Should we have similar EVs in the project level?

Yes. Sorry if that wasn't clear in the issue description, but it's why I reference the launcher & device separately. The expectation is PROJECT_NAME/ID exist wherever the project is running - in the platform or on a device.

It is also why I separated out Project and Device EVs in my previous reply. The project will provide EVs specific to the project, the device will add in EVs specific to it. This is why project-specific EVS shouldn't be stored on the device itself - they come from the project.

**Comment by Steve-Mcl** [comment]
> For the purposes of this isue - getting the basic infrastructure in place to have these platform provided EVs, can we stick to the ones listed in the issue?

No bother, makes sense.


Gotcha on the other stuff too. Should get this done tomorrow now with a clearer path.

**Comment by Steve-Mcl** [comment]
ok, after much ado - this is the current operational state...

![image](https://user-images.githubusercontent.com/44235289/194330413-ba09cd82-4550-4372-ac43-5c447e31976f.png)

Demo flow (for reference and manual testing) ...

```
[{"id":"a0e887a9c261f1ef","type":"http in","z":"171e70b2de3c4bb7","name":"","url":"/whoami","method":"get","upload":false,"swaggerDoc":"","x":170,"y":100,"wires":[["c81d5f661cc49e7d"]]},{"id":"c81d5f661cc49e7d","type":"function","z":"171e70b2de3c4bb7","name":"env.get('name')","func":"\nconst m = RED.util.cloneMessage(msg);\ndelete m.res\ndelete m.req\nm.payload = {}\nconst p = m.payload;\n\np.templateDefined = {\n    'template-name': env.get('template-name'),\n    'instance-name': env.get('instance-name'),\n}\n\np.projectDefined = {\n    ['proj-var']: env.get('proj-var'),\n    FF_PROJECT_ID: env.get('FF_PROJECT_ID'),\n    FF_PROJECT_NAME: env.get('FF_PROJECT_NAME'),\n}\n\np.deviceDefined = {\n    ['device-var']: env.get('device-var'),\n    FF_DEVICE_ID: env.get('FF_DEVICE_ID'),\n    FF_DEVICE_NAME: env.get('FF_DEVICE_NAME'),\n    FF_DEVICE_TYPE: env.get('FF_DEVICE_TYPE')\n}\n\nif(msg.noHttpResponse === true) {\n    return [null, m];    \n}\nreturn [msg, m];","outputs":2,"noerr":0,"initialize":"","finalize":"","libs":[],"x":400,"y":100,"wires":[["292d789a5690689c"],["163e3acd4e1028cc"]]},{"id":"292d789a5690689c","type":"http response","z":"171e70b2de3c4bb7","name":"","statusCode":"","headers":{},"x":650,"y":100,"wires":[]},{"id":"163e3acd4e1028cc","type":"project link out","z":"171e70b2de3c4bb7","name":"","mode":"link","broadcast":true,"project":"ca65f5ed-aea0-4a10-ac9a-2086b6af6700","topic":"whoami_query","x":620,"y":160,"wires":[]},{"id":"b3b854b50f730037","type":"debug","z":"171e70b2de3c4bb7","name":"Check full msg - see if env vars made it through","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":500,"y":220,"wires":[]},{"id":"8eeecd437a556fb8","type":"project link in","z":"171e70b2de3c4bb7","name":"whoami_query","project":"all","broadcast":true,"topic":"whoami_query","x":180,"y":220,"wires":[["b3b854b50f730037"]]},{"id":"43c3a99dfbe4ed05","type":"inject","z":"171e70b2de3c4bb7","name":"","props":[{"p":"noHttpResponse","v":"true","vt":"bool"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":150,"y":160,"wires":[["c81d5f661cc49e7d"]]},{"id":"74075860c12e9448","type":"comment","z":"171e70b2de3c4bb7","name":"Enpoint to trigger flow from Project or from remote Device","info":"","x":310,"y":60,"wires":[]}]
```

Will push latest shortly. 

If ok, I will develop tests.

**Comment by Steve-Mcl** [comment]
changes pushed and tests added.

**Comment by Steve-Mcl** [comment]
@hardillb @knolleary 

Need more heads on this one pls.

Ben has confirmed everything works but he found a gap where the new env vars are not sent to an existing device.

The settings hash is stored in the DB table (to minimise recalcs) but on  existing device(s), unless the user addd a new env-var to the device, the call to recalculate the settings hash is not made & therefore not saved in the table.

What would be a sensible approach to this? 
* Compute settings hash on the fly?
* A DB migration?
* Something else?

Thanks.

**Comment by hardillb** [comment]
Since we are adding new env vars as a basically a one off action for existing devices with this deploy, a db migration that triggers a rehash feels right

**Comment by Steve-Mcl** [comment]
confirmed variables are present in UI


#### project


![Image](https://user-images.githubusercontent.com/44235289/197711222-8055c63a-f8ca-4cbc-bfd1-dd1be87b3ddb.png)




#### devices


![Image](https://user-images.githubusercontent.com/44235289/197711024-26f3697f-9f63-49f2-b88a-3153468dea23.png)

**Comment by Steve-Mcl** [comment]
verified locally (lastest main commit as of 20 mins ago)
verified on cloud ~ local device
